flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sink/ConstraintValidator.java

```diff
@@ -0,0 +1,201 @@
+            if (rowData.isNullAt(index)) {
+                switch (notNullEnforcer) {
+                    case ERROR:
+                        throw new TableException(
+                                String.format(
+                                        "Column '%s' is NOT NULL, however, a null value is being written into it. "
+                                                + "You can set job configuration '%s'='drop' "
+                                                + "to suppress this exception and drop such records silently.",
+                                        allFieldNames[index],
+                                        TABLE_EXEC_SINK_NOT_NULL_ENFORCER.key()));
+                    case DROP:
+                        return null;
+                }
+            }
+        }
+        return element;
+    }
+
+    private StreamRecord<RowData> processCharConstraint(StreamRecord<RowData> element) {
+        if (charPrecisionEnforcer == null) {
+            return element;
+        }
+
+        final RowData rowData = element.getValue();
+        final List<Tuple2<Integer, String>> newStrings = new ArrayList<>();
+        for (Tuple2<Integer, Integer> tuple : charFields) {
+            final int fieldIdx = tuple.f0;
+            final int length = tuple.f1;
+            final String stringValue = rowData.getString(fieldIdx).toString();
+
+            if (stringValue.length() > length) {
+                switch (charPrecisionEnforcer) {
+                    case ERROR:
+                        throw new TableException(
+                                String.format(
+                                        "Column '%s' is of CHAR/VARCHAR type with a length of '%s', however, "
+                                                + "a string value with a length of '%s' is being written into it. "
+                                                + "You can set job configuration '%s'='trim' to suppress this exception "
+                                                + "and trim such string values to the length defined by the CHAR/VARCHAR "
+                                                + "type.",
+                                        allFieldNames[fieldIdx],
+                                        length,
+                                        stringValue.length(),
+                                        TABLE_EXEC_SINK_CHAR_PRECISION_ENFORCER.key()));
+                    case TRIM:
+                        newStrings.add(new Tuple2<>(fieldIdx, stringValue.substring(0, length)));
+                        break;
+                }
+            }
+        }
+
+        if (newStrings.isEmpty()) {
+            return element;
+        }
+
+        UpdatableRowData newRowData = new UpdatableRowData(rowData, rowData.getArity());
+        for (Tuple2<Integer, String> tuple : newStrings) {
+            newRowData.setField(tuple.f0, StringData.fromString(tuple.f1));
+        }
+        return new StreamRecord<>(newRowData);
```

Comment:
can we reuse `StreamRecord`? Object creation should be avoided in the hot path. How do other operators do this?

Reply:
yes, with `replace`